File: algorithms/285.Inorder_Successor_in_BST/Perfect_Squares.py
<<<<<<< HEAD
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderSuccessor(self, root, p):
        """
        :type root: TreeNode
        :type p: TreeNode
        :rtype: TreeNode
        """
        # basically inorder travesal return the one next to p
        prev = None
        path = [root]
        while path:
            cur = path[-1]
            if cur:
                path.append(cur.left)
            else:
                path.pop(-1)
                if path:
                    cur = path.pop(-1)
                    if prev == p:
                        return cur
                    else:
                        prev = cur
                    path.append(cur.right)
        return None
=======
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def inorderSuccessor(self, root, p):
        """
        :type root: TreeNode
        :type p: TreeNode
        :rtype: TreeNode
        """
        # basically inorder travesal return the one next to p
        prev = None
        path = [root]
        while path:
            cur = path[-1]
            if cur:
                path.append(cur.left)
            else:
                path.pop(-1)
                if path:
                    cur = path.pop(-1)
                    if prev == p:
                        return cur
                    else:
                        prev = cur
                    path.append(cur.right)
        return None
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74

File: algorithms/285.Inorder_Successor_in_BST/Perfect_Squares.cc
<<<<<<< HEAD
// Inorder Successor in BST
class Solution {
public:
  TreeNode* inorderSuccessor(TreeNode* x, TreeNode* p) {
    TreeNode *r = NULL;
    while (x) {
      if (p->val < x->val) {
        r = x;
        x = x->left;
      } else
        x = x->right;
    }
    return r;
  }
=======
// Inorder Successor in BST
class Solution {
public:
  TreeNode* inorderSuccessor(TreeNode* x, TreeNode* p) {
    TreeNode *r = NULL;
    while (x) {
      if (p->val < x->val) {
        r = x;
        x = x->left;
      } else
        x = x->right;
    }
    return r;
  }
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
};